core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -910,7 +910,7 @@ protected Scope visitCreateTableAsSelect(CreateTableAsSelect node, Optional<Scop
             }
 
             // create target table metadata
-            ConnectorTableMetadata tableMetadata = new ConnectorTableMetadata(targetTable.asSchemaTableName(), columns.build(), properties, node.getComment());
+            ConnectorTableMetadata tableMetadata = metadata.getNewTableEffectiveMetadata(session, catalogName, targetTable, columns.build(), properties, node.getComment());
```

Comment:
Also, we must decide on the level of flexibility: I think the minimum plan would be to support implicit coercions only, but I'm not sure if it covers the targeted cases. And if we go beyond that minimum, do we want to fail on string truncation, etc.?
When we have the decision, then we need to verify if the new type returned by the connector satisfies our criteria.

The method `typesMatchForInsert()` can serve as a reference, but please note that it reflects (with some limitations) what the SQL standard says about inserts. The CTAS case is different, and it is not covered by the spec.

Reply:
> Also, we must decide on the level of flexibility: I think the minimum plan would be to support implicit coercions only, but I'm not sure if it covers the targeted cases.

Today, eg PostgreSQL (and co.) connector accepts any `timestamp(n)` and coerces it to `timestamp(6)`.
This all happens inside the connector (in every such connector),  and is an unnecessary complexity, since the coercion logic is effectively spread out.

For `n <= 6` an implicit coercion would do, but for `n > 6` we need to allow non-implicit coercions.